app/questionnaire/value_source_resolver.py

```diff
@@ -0,0 +1,124 @@
+from dataclasses import dataclass
+from typing import Optional, Union
+
+from app.data_models import AnswerStore, ListStore
+from app.data_models.list_store import ListModel
+from app.questionnaire import Location, QuestionnaireSchema
+from app.questionnaire.relationship_location import RelationshipLocation
+
+answer_value_types = Union[str, int, float, list, dict, None]
+value_source_types = Union[str, int, float, list, None]
+
+
+@dataclass
+class ValueSourceResolver:
+    answer_store: AnswerStore
+    list_store: ListStore
+    metadata: dict
+    schema: QuestionnaireSchema
+    location: Union[Location, RelationshipLocation]
+    list_item_id: Optional[str]
+    routing_path_block_ids: Optional[list] = None
+    use_default_answer: bool = False
+
+    def _get_list_item_id_from_value_source(self, value_source: dict) -> Optional[str]:
+        if not (list_item_selector := value_source.get("list_item_selector")):
+            return None
+
+        value: Optional[str] = None
+        if list_item_selector["source"] == "location":
+            value = getattr(self.location, list_item_selector["id"])
+
+        elif list_item_selector["source"] == "list":
+            value = getattr(
+                self.list_store[list_item_selector["id"]],
+                list_item_selector["id_selector"],
+            )
+
+        return value
+
+    def _is_answer_on_path(self, answer_id: str) -> bool:
+        if self.routing_path_block_ids:
+            block = self.schema.get_block_for_answer_id(answer_id)
+            return block is not None and block["id"] in self.routing_path_block_ids
+
+        return True
+
+    def _get_answer_value(
+        self, answer_id: str, list_item_id: Optional[str]
+    ) -> answer_value_types:
+        if not self._is_answer_on_path(answer_id):
+            return None
+
+        if answer := self.answer_store.get_answer(answer_id, list_item_id):
+            return answer.value
+
+        if self.use_default_answer and (
+            answer := self.schema.get_default_answer(answer_id)
+        ):
+            return answer.value
+
+    def _resolve_answer_value(self, value_source: dict) -> value_source_types:
```

Comment:
This is essentially because of the first point in https://github.com/ONSdigital/eq-questionnaire-runner/pull/630#discussion_r676646236.
Can rename the value to `resolved_answer_value` but I don't think it adds much.